packages/iov-core/src/jsonrpcsigningserver.spec.ts

```diff
@@ -0,0 +1,260 @@
+  });
+
+  it("can get BNS or Ethereum identities", async () => {
+    pendingWithoutBnsd();
+    pendingWithoutEthereum();
+
+    const bnsConnection = await bnsConnector(bnsdUrl).client();
+
+    const server = await makeBnsEthereumSigningServer();
+
+    const response = await server.handleChecked({
+      jsonrpc: "2.0",
+      id: 123,
+      method: "getIdentities",
+      params: {
+        reason: "Who are you?",
+        chainIds: [ethereumChainId, bnsConnection.chainId()],
+      },
+    });
+    expect(response.id).toEqual(123);
+    if (isJsonRpcErrorResponse(response)) {
+      throw new Error(`Response must not be an error, but got '${response.error.message}'`);
+    }
+    const result = TransactionEncoder.fromJson(response.result);
+    expect(result).toEqual(jasmine.any(Array));
+    expect((result as ReadonlyArray<any>).length).toEqual(2);
+    expect(result[0].chainId).toEqual(bnsConnection.chainId());
+    expect(result[0].pubkey).toEqual({
+      algo: Algorithm.Ed25519,
+      data: fromHex("533e376559fa551130e721735af5e7c9fcd8869ddd54519ee779fce5984d7898"),
+    });
+    expect(result[1]).toEqual(ganacheSecondIdentity);
+
+    server.shutdown();
+    bnsConnection.disconnect();
+  });
+
+  it("send a signing request to service", async () => {
+    pendingWithoutBnsd();
+    pendingWithoutEthereum();
+
+    const bnsConnection = await bnsConnector(bnsdUrl).client();
+
+    const server = await makeBnsEthereumSigningServer();
+
+    const identitiesResponse = await server.handleChecked({
+      jsonrpc: "2.0",
+      id: 1,
+      method: "getIdentities",
+      params: {
+        reason: "Who are you?",
+        chainIds: [bnsConnection.chainId()],
+      },
+    });
+    if (isJsonRpcErrorResponse(identitiesResponse)) {
+      throw new Error(`Response must not be an error, but got '${identitiesResponse.error.message}'`);
+    }
+    expect(identitiesResponse.result).toEqual(jasmine.any(Array));
+    expect((identitiesResponse.result as ReadonlyArray<any>).length).toEqual(1);
+    const signer = TransactionEncoder.fromJson(identitiesResponse.result[0]);
+    if (!isPublicIdentity(signer)) {
+      throw new Error("Identity element is not valid");
+    }
+
+    const send: SendTransaction = {
+      kind: "bcp/send",
+      creator: signer,
+      memo: `Hello ${Math.random()}`,
+      amount: defaultAmount,
+      recipient: await randomBnsAddress(),
+    };
+
+    const signAndPostResponse = await server.handleChecked({
+      jsonrpc: "2.0",
+      id: 2,
+      method: "signAndPost",
+      params: {
+        reason: "Please sign",
+        transaction: TransactionEncoder.toJson(send),
+      },
+    });
+    if (isJsonRpcErrorResponse(signAndPostResponse)) {
+      throw new Error(`Response must not be an error, but got '${signAndPostResponse.error.message}'`);
+    }
+    const transactionId: TransactionId = TransactionEncoder.fromJson(signAndPostResponse.result);
+    expect(transactionId).toMatch(/^[0-9A-F]+$/);
```

Comment:
Because the BCP implementation of BNS guarantees to use upper case hex for transaction IDs. This property is strictly required to allow string equality on BCP's `TransactionId`s.